src/components/form/ha-form.ts

```diff
@@ -0,0 +1,179 @@
+import {
+  customElement,
+  LitElement,
+  html,
+  property,
+  query,
+  CSSResult,
+  css,
+  PropertyValues,
+} from "lit-element";
+
+import "./ha-form-string";
+import "./ha-form-integer";
+import "./ha-form-float";
+import "./ha-form-boolean";
+import "./ha-form-select";
+import { fireEvent } from "../../common/dom/fire_event";
+
+export interface HaFormSchema {
+  type: "string" | "integer" | "float" | "boolean" | "select";
+  required: boolean;
+  name: string;
+  valueMin?: number;
+  valueMax?: number;
+  options?: string[];
+  description?: { unit_of_measurement?: string };
+}
+
+export interface HaFormData {
+  [key: string]: any;
+}
+
+export interface HaFormElement extends LitElement {
+  schema: HaFormSchema;
+  data: HaFormData;
+  label?: string;
+  suffix?: string;
+}
+
+@customElement("ha-form")
+export class HaForm extends LitElement implements HaFormElement {
+  @property() public data!: HaFormData;
+  @property() public schema!: HaFormSchema;
+  @property() public error;
+  @property() public computeError?: (schema: HaFormSchema, error) => string;
+  @property() public computeLabel?: (schema: HaFormSchema) => string;
+  @property() public computeSuffix?: (schema: HaFormSchema) => string;
+  @query("ha-form") private _childForm?: HaForm;
+  @query("#element") private _elementContainer?: HTMLDivElement;
+
+  public focus() {
+    const input = this._childForm
+      ? this._childForm
+      : this._elementContainer
+      ? this._elementContainer.lastChild
+      : undefined;
+
+    if (!input) {
+      return;
+    }
+
+    (input as HTMLElement).focus();
+  }
+
+  protected render() {
+    if (Array.isArray(this.schema)) {
+      return html`
+        ${this.error && this.error.base
+          ? html`
+              <div class="error">
+                ${this._computeError(this.error.base, this.schema)}
+              </div>
+            `
+          : ""}
+        ${this.schema.map(
+          (item) => html`
+            <ha-form
+              .data=${this._getValue(this.data, item)}
+              .schema=${item}
+              .error=${this._getValue(this.error, item)}
+              @value-changed=${this._valueChanged}
+              .computeError=${this.computeError}
+              .computeLabel=${this.computeLabel}
+              .computeSuffix=${this.computeSuffix}
+            ></ha-form>
+          `
+        )}
+      `;
+    }
+
+    return html`
+      ${this.error
+        ? html`
+            <div class="error">
+              ${this._computeError(this.error, this.schema)}
+            </div>
+          `
+        : ""}
+      <div id="element" @value-changed=${this._valueChanged}></div>
+    `;
+  }
+
+  protected updated(changedProperties: PropertyValues) {
+    const oldSchema: HaFormSchema = changedProperties.get(
+      "schema"
+    ) as HaFormSchema;
+    if (
+      !Array.isArray(this.schema) &&
```

Comment:
What if the current schema is an array, shouldn't we clean up the element ?

Reply:
Render takes care of that, `div#element` is only rendered if the schema is not an array